osfoffline/utils/__init__.py

```diff
@@ -37,6 +39,38 @@ def hash_file(path, *, chunk_size=65536):
     return s.hexdigest()
 
 
+def legal_filename(basename, *, parent=None):
+    """
+    Replace all OS-illegal characters in a filename with underscore, and return a new fn guaranteed to be unique
+        in that folder
+    :param str basename: The basename of a file, without path (eg 'is this a file and/or blob?.txt')
+    :param models.File parent: If provided, will verify that the new aliased name is unique in the
+        parent folder (or project osf storage folder). Two aliased names should not collide.
+    :return:
+    """
+    # TODO: After we handle filenames, explore whether project names can also have illegal characters
+    # TODO: Add a unit test!
+    if sys.platform in ('win32', 'cygwin'):
+        # https://support.microsoft.com/en-us/kb/177506
+        illegal_chars = r'\/:*?"<>|'
+        new_fn = re.sub(r'[{}]'.format(illegal_chars), '_', basename)
+        n = 1
+        while parent:
+            # If parent node is provided, loop through until a valid filename (not in use) is available
+            if Session().query(models.File.name).filter(models.File.alias == new_fn,
```

Comment:
Could be better to create a set of taken filenames outside of this loop rather than re-running the query at each iteration.